fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexPrefetchRangeKeyValueCursor.java

```diff
@@ -0,0 +1,84 @@
+import com.apple.foundationdb.MappedKeyValue;
+import com.apple.foundationdb.ReadTransaction;
+import com.apple.foundationdb.StreamingMode;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.AsyncIterator;
+import com.apple.foundationdb.record.cursors.CursorLimitManager;
+import com.apple.foundationdb.subspace.Subspace;
+
+import javax.annotation.Nonnull;
+
+/**
+ * A {@link KeyValueCursor} that scans an index using the IndexPrefetch operation.
+ * This subclass of {@link KeyValueCursor} uses a different scan operation ({@link com.apple.foundationdb.Transaction#getMappedRange})
+ * and therefore the actual returned types of the scanned range are different too - they need to be parsed into records
+ * rather than Index entries. The returned values are MappedKeyValues and contain all splits of
+ * the record for a particular index entry.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class IndexPrefetchRangeKeyValueCursor extends KeyValueCursor {
+    private IndexPrefetchRangeKeyValueCursor(@Nonnull final FDBRecordContext context,
+                                             @Nonnull final AsyncIterator<KeyValue> iterator,
+                                             int prefixLength,
+                                             @Nonnull final CursorLimitManager limitManager,
+                                             int valuesLimit) {
+
+        super(context, iterator, prefixLength, limitManager, valuesLimit);
+    }
+
+    /**
+     * A Builder for the cursor. Note that this builder actually extends the superclass' builder, and does not create
+     * a {@link IndexPrefetchRangeKeyValueCursor} but rather a {@link KeyValueCursor}, by virtue of the fact that it
+     * only extends teh {@link #scanRange} method. This is by design and the created cursor is a {@link KeyValueCursor}
+     * with no behavior differences from the default.
+     */
+    @API(API.Status.EXPERIMENTAL)
+    public static class Builder extends KeyValueCursor.Builder {
+        // The HopInfo that is used for the getRangeAndFlatMap call
+        private final byte[] hopInfo;
+
+        private Builder(@Nonnull Subspace indexSubspace, @Nonnull byte[] hopInfo) {
+            super(indexSubspace);
+            this.hopInfo = hopInfo;
+        }
+
+        public static Builder newBuilder(@Nonnull Subspace indexSubspace, @Nonnull byte[] hopInfo) {
+            return new Builder(indexSubspace, hopInfo);
+        }
+
+        @Override
+        protected AsyncIterator<MappedKeyValue> scanRange(@Nonnull ReadTransaction transaction,
+                                                          @Nonnull KeySelector begin,
+                                                          @Nonnull KeySelector end,
+                                                          int limit, boolean reverse,
+                                                          @Nonnull StreamingMode streamingMode) {
+            return transaction
```

Comment:
Should this validate the API version somehow as a safeguard? I think your checks in the index plan are sufficient as long as people go through the planner, but I'm thinking of protecting ourselves in case someone calls `recordStore.scanIndexPrefetch` directly, and this spot seemed like the most direct (though I could maybe see it being checked in `recordStore.scanIndexPrefetch` or something instead, if you think that's better)

Reply:
Done. Added a check in the store's `scanIndexRemoteFetchInternal` method.